math/src/main/kotlin/ktx/math/ImmutableVector.kt

```diff
@@ -0,0 +1,154 @@
+    operator fun plus(other: T): T
+
+    /** Returns the same vector with all members incremented by 1 */
+    operator fun inc(): T
+
+    /** Returns the same vector with all members decremented by 1 */
+    operator fun dec(): T
+
+    /** Returns this vector scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** Returns his vector scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** Returns he dot product of this vector by the given [vector] */
+    infix fun dot(vector: T): Float
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    infix fun dst2(vector: T): Float
+
+    /** Linearly interpolates between this vector and the target vector by alpha */
+    fun lerp(target: T, alpha: Float): T
+
+    /** Returns a vector of same length and a random direction */
+    fun withRandomDirection(): T
+
+    /** Returns true if this vector is in line with the other vector (either in the same or the opposite direction) */
+    fun isOnLine(other: T, epsilon: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Compares this vector with the other vector, using the supplied epsilon for fuzzy equality testing */
+    fun epsilonEquals(other: T, epsilon: Float): Boolean
+
+    override fun compareTo(other: T): Int = len2.compareTo(other.len2)
+}
+
+/**
+ * Returns the euclidean length
+ */
+inline val ImmutableVector<*>.len: Float get() = sqrt(len2)
+
+/**
+ * Returns the unit vector of same direction or this vector if it is zero.
+ */
+val <T : ImmutableVector<T>> T.nor: T
+    get() {
+        val l2 = len2
+
+        if (l2 == 1f || l2 == 0f) return this
+
+        return withLength2(1f)
+    }
+
+/** Returns a vector of the same direction and a squared length of [length2] */
+fun <T : ImmutableVector<T>> T.withLength2(length2: Float): T {
+    val oldLen2 = this.len2
+
+    if (oldLen2 == 0f || oldLen2 == length2) return this
+
+    return times(sqrt(length2 / oldLen2))
+}
+
+/** Returns whether this vector is a unit length vector within the given [margin]. (no margin by default) */
+inline fun <T : ImmutableVector<T>> T.isUnit(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean = abs(1f - len2) < margin
+
+/** Returns the distance between this and the [other] vector */
+inline infix fun <T : ImmutableVector<T>> T.dst(other: T): Float = sqrt(dst2(other))
+
+/** Returns this vector scaled by (1 / [scalar]) */
+inline operator fun <T : ImmutableVector<T>> T.div(scalar: Float): T = times(1 / scalar)
+
+/** Returns this vector if the [ImmutableVector.len] is <= [limit] or a vector with the same direction and length [limit] otherwise */
+inline fun <T : ImmutableVector<T>> T.limit(limit: Float): T = limit2(limit * limit)
```

Comment:
Not sure if most of these methods should be inlined and outside of the interface. As I've mentioned, we usually inline functions with lambda parameters or reified generics. Is there a reason you inlined most of this outside of reducing method count?

Reply:
Because I consider them aliases. I don't consider them to have their own proper logic. Therefore the main problem of inline function which is the fact they expose their implementation, does not apply.

But as I said in my other comments, I'll remove the inline modifier, if you prefer.